src/jbmc/jbmc_parse_options.cpp

```diff
@@ -0,0 +1,922 @@
+    if(!version_set)
+      options.set_option("smt2", true), version_set=true;
+  }
+
+  if(cmdline.isset("yices"))
+  {
+    options.set_option("yices", true), solver_set=true;
+    if(!version_set)
+      options.set_option("smt2", true), version_set=true;
+  }
+
+  if(cmdline.isset("z3"))
+  {
+    options.set_option("z3", true), solver_set=true;
+    if(!version_set)
+      options.set_option("smt2", true), version_set=true;
+  }
+
+  if(cmdline.isset("opensmt"))
+  {
+    options.set_option("opensmt", true), solver_set=true;
+    if(!version_set)
+      options.set_option("smt1", true), version_set=true;
+  }
+
+  if(version_set && !solver_set)
+  {
+    if(cmdline.isset("outfile"))
+    {
+      // outfile and no solver should give standard compliant SMT-LIB
+      options.set_option("generic", true), solver_set=true;
+    }
+    else
+    {
+      if(options.get_bool_option("smt1"))
+      {
+        options.set_option("boolector", true), solver_set=true;
+      }
+      else
+      {
+        assert(options.get_bool_option("smt2"));
+        options.set_option("z3", true), solver_set=true;
+      }
+    }
+  }
+  // Either have solver and standard version set, or neither.
+  assert(version_set==solver_set);
+
+  if(cmdline.isset("beautify"))
+    options.set_option("beautify", true);
+
+  if(cmdline.isset("no-sat-preprocessor"))
+    options.set_option("sat-preprocessor", false);
+  else
+    options.set_option("sat-preprocessor", true);
+
+  options.set_option(
+    "pretty-names",
+    !cmdline.isset("no-pretty-names"));
+
+  if(cmdline.isset("outfile"))
+    options.set_option("outfile", cmdline.get_value("outfile"));
+
+  if(cmdline.isset("graphml-witness"))
+  {
+    options.set_option("graphml-witness", cmdline.get_value("graphml-witness"));
+    options.set_option("stop-on-fail", true);
+    options.set_option("trace", true);
+  }
+
+  if(cmdline.isset("symex-coverage-report"))
+    options.set_option(
+      "symex-coverage-report",
+      cmdline.get_value("symex-coverage-report"));
+}
+
+/// invoke main modules
+int jbmc_parse_optionst::doit()
+{
+  if(cmdline.isset("version"))
+  {
+    std::cout << CBMC_VERSION << '\n';
+    return 0; // should contemplate EX_OK from sysexits.h
+  }
+
+  //
+  // command line options
+  //
+
+  optionst options;
+  try
+  {
+    get_command_line_options(options);
+  }
+
+  catch(const char *error_msg)
+  {
+    error() << error_msg << eom;
+    return 6; // should contemplate EX_SOFTWARE from sysexits.h
+  }
+
+  catch(const std::string error_msg)
+  {
+    error() << error_msg << eom;
+    return 6; // should contemplate EX_SOFTWARE from sysexits.h
+  }
+
+  eval_verbosity();
+
+  //
+  // Print a banner
+  //
+  status() << "JBMC version " CBMC_VERSION " "
+           << sizeof(void *)*8 << "-bit "
+           << config.this_architecture() << " "
+           << config.this_operating_system() << eom;
+
+  register_language(new_ansi_c_language);
```

Comment:
Why register C? If it's purely for a fallback for `from_expr` et al suggest comment to this effect.

Reply:
No, we actually do add C-mode symbols (rounding mode, etc.)